File: R/idw.R
#
#  idw.R
#
#  Inverse-distance weighted smoothing
#
#  $Revision: 1.11 $ $Date: 2018/12/16 03:49:58 $

idw <- function(X, power=2, at=c("pixels", "points"), ..., se=FALSE) {
  stopifnot(is.ppp(X) && is.marked(X))
  at <- match.arg(at)
  marx <- marks(X)
  if(is.data.frame(marx)) {
    if((nc <- ncol(marx)) > 1) {
      ## multiple columns of marks - process one-by-one
      each <- vector(mode="list", length=nc)
      for(j in 1:nc) 
        each[[j]] <- idw(X %mark% marx[,j], power=power, at=at, ..., se=se)
      names(each) <- colnames(marx)
      ##
      if(!se) {
        ## estimates only
        switch(at,
               pixels = { out <- as.solist(each) },
               points = { out <- as.data.frame(each) }
               )
      } else {
        ## estimates and standard errors
        est <- lapply(each, getElement, name="estimate")
        SE  <- lapply(each, getElement, name="SE")
        switch(at,
               pixels = {
                 out <- list(estimate = as.solist(est),
                             SE       = as.solist(SE))
               },
               points = {
                 out <- list(estimate = as.data.frame(est),
                              SE      = as.data.frame(SE))
               })
      }
      return(out)
    } else 
      marx <- marx[,1L]
  }
  if(!is.numeric(marx))
    stop("Marks must be numeric")
  check.1.real(power)
  switch(at,
         pixels = {
           ## create grid
           W <- as.mask(as.owin(X), ...)
           dim <- W$dim
           npixels <- prod(dim)
           ## call C
           if(!se) {
             z <- .C("Cidw",
                     x = as.double(X$x),
                     y = as.double(X$y),
                     v = as.double(marx),
                     n = as.integer(npoints(X)),
                     xstart = as.double(W$xcol[1L]),
                     xstep  = as.double(W$xstep),
                     nx     = as.integer(dim[2L]),
                     ystart = as.double(W$yrow[1L]),
                     ystep  = as.double(W$ystep),
                     ny     = as.integer(dim[1L]),
                     power  = as.double(power),
                     num    = as.double(numeric(npixels)),
                     den    = as.double(numeric(npixels)),
                     rat    = as.double(numeric(npixels)),
                     PACKAGE = "spatstat")
             out <- as.im(matrix(z$rat, dim[1L], dim[2L]), W=W)
             out <- out[W, drop=FALSE]
           } else {
             z <- .C("Cidw2",
                     x = as.double(X$x),
                     y = as.double(X$y),
                     v = as.double(marx),
                     n = as.integer(npoints(X)),
                     xstart = as.double(W$xcol[1L]),
                     xstep  = as.double(W$xstep),
                     nx     = as.integer(dim[2L]),
                     ystart = as.double(W$yrow[1L]),
                     ystep  = as.double(W$ystep),
                     ny     = as.integer(dim[1L]),
                     power  = as.double(power),
                     num    = as.double(numeric(npixels)),
                     den    = as.double(numeric(npixels)),
                     rat    = as.double(numeric(npixels)),
                     mtwo   = as.double(numeric(npixels)),
                     wtwo   = as.double(numeric(npixels)),
                     PACKAGE = "spatstat")
             est <- as.im(matrix(z$rat, dim[1L], dim[2L]), W=W)
             est <- est[W, drop=FALSE]
             sumw  <- z$den
             sumw2 <- z$wtwo
             m2    <- z$mtwo
             varden <- sumw - sumw2/sumw
             varden[varden <= 0] <- NA
             SE <- sqrt(m2/varden)
             SE <- as.im(matrix(SE, dim[1L], dim[2L]), W=W)
             SE <- SE[W, drop=FALSE]
             out <- list(estimate=est, SE=SE)
           }
         },
         points={
           npts <- npoints(X)
           if(!se) {
             z <- .C("idwloo",
                     x = as.double(X$x),
                     y = as.double(X$y),
                     v = as.double(marx),
                     n = as.integer(npts),
                     power  = as.double(power),
                     num    = as.double(numeric(npts)),
                     den    = as.double(numeric(npts)),
                     rat    = as.double(numeric(npts)),
                     PACKAGE = "spatstat")
             out <- z$rat
           } else {
             z <- .C("idwloo2",
                     x = as.double(X$x),
                     y = as.double(X$y),
                     v = as.double(marx),
                     n = as.integer(npts),
                     power  = as.double(power),
                     num    = as.double(numeric(npts)),
                     den    = as.double(numeric(npts)),
                     rat    = as.double(numeric(npts)),
                     mtwo   = as.double(numeric(npts)),
                     wtwo   = as.double(numeric(npts)),
                     PACKAGE = "spatstat")
             est <- z$rat
             sumw  <- z$den
             sumw2 <- z$wtwo
             m2    <- z$mtwo
             varden <- sumw - sumw2/sumw
             varden[varden <= 0] <- NA
             SE <- sqrt(m2/varden)
             out <- list(estimate=est, SE=SE)
           }
         })
  return(out)
}

File: src/idw.c
/*
  idw.c

  Inverse-distance weighted smoothing

  $Revision: 1.11 $ $Date: 2018/12/18 02:43:11 $

  Cidw    inverse distance smoothing from data points onto pixel grid
  idwloo  leave-one-out estimate at data points

  Cidw2   Cidw with variance estimate
  idwloo2 idwloo with variance estimate

  Copyright (C) Adrian Baddeley, Ege Rubak and Rolf Turner 2001-2018
  Licence: GNU Public Licence >= 2

*/

#include <Rmath.h>
#include <R_ext/Utils.h>
#include "chunkloop.h"

#define MAT(X,I,J,NROW) (X)[(J) + (NROW) * (I)]

/*  inverse-distance smoothing from data points onto pixel grid */

void Cidw(x, y, v, n, xstart, xstep, nx, ystart, ystep, ny, power, num, den, rat)
     double *x, *y, *v;           /* data points and values */
     int *n;
     double *xstart, *xstep, *ystart, *ystep;   /* pixel grid */
     int *nx, *ny;
     double *power;                   /* exponent for IDW */
     double *num, *den, *rat;     /* output arrays - assumed initialised 0 */
{
  int N, i, Nx, Ny, ix, iy;
  double xg, yg, x0, dx, y0, dy, pon2, d2, w, sumw, sumwv;
  
  N  = *n;
  Nx = *nx;
  Ny = *ny;
  x0 = *xstart;
  y0 = *ystart;
  dx = *xstep;
  dy = *ystep;

  pon2 = (*power)/2.0;

  if(pon2 == 1.0) {
    /* slightly faster code when power=2 */
    for(ix = 0, xg=x0; ix < Nx; ix++, xg+=dx) {
      if(ix % 256 == 0) R_CheckUserInterrupt();
      for(iy = 0, yg=y0; iy < Ny; iy++, yg+=dy) {
	sumwv = sumw = 0.0;
	/* loop over data points, accumulating numerator and denominator */
	for(i = 0; i < N; i++) {
	  d2 = (xg - x[i]) * (xg - x[i]) + (yg - y[i]) * (yg - y[i]);
	  w = 1.0/d2;
	  sumwv += w * v[i];
	  sumw  += w;
	}
	/* compute ratio */
	MAT(num, ix, iy, Ny) = sumwv;
	MAT(den, ix, iy, Ny) = sumw;
	MAT(rat, ix, iy, Ny) = sumwv/sumw;
      }
    }
  } else {
    /* general case */
    for(ix = 0, xg=x0; ix < Nx; ix++, xg+=dx) {
      if(ix % 256 == 0) R_CheckUserInterrupt();
      for(iy = 0, yg=y0; iy < Ny; iy++, yg+=dy) {
	sumwv = sumw = 0.0;
	/* loop over data points, accumulating numerator and denominator */
	for(i = 0; i < N; i++) {
	  d2 = (xg - x[i]) * (xg - x[i]) + (yg - y[i]) * (yg - y[i]);
	  w = 1.0/pow(d2, pon2);
	  sumwv += w * v[i];
	  sumw  += w;
	}
	/* compute ratio */
	MAT(num, ix, iy, Ny) = sumwv;
	MAT(den, ix, iy, Ny) = sumw;
	MAT(rat, ix, iy, Ny) = sumwv/sumw;
      }
    }
  }
}

/* Leave-one-out IDW at data points only */

void idwloo(x, y, v, n, power, num, den, rat)
     double *x, *y, *v;           /* data points and values */
     int *n;
     double *power;                   /* exponent for IDW */
     double *num, *den, *rat;     /* output vectors - assumed initialised 0 */
{
  int N, i, j, maxchunk;
  double xi, yi, d2, w, pon2, sumw, sumwv;
  
  N  = *n;
  pon2 = (*power)/2.0;

  if(pon2 == 1.0) {
    /* slightly faster code when power=2 */
    OUTERCHUNKLOOP(i, N, maxchunk, 16384) {
      R_CheckUserInterrupt();
      INNERCHUNKLOOP(i, N, maxchunk, 16384) {
	xi = x[i];
	yi = y[i];
	sumwv = sumw = 0.0;
	if(i > 0) {
	  for(j = 0; j < i; j++) {
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/d2;
	    sumwv += w * v[j];
	    sumw  += w;
	  }
	}
	if(i < N-1) {
	  for(j = i+1; j < N; j++) {
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/d2;
	    sumwv += w * v[j];
	    sumw  += w;
	  }
	}
	/* compute ratio */
	num[i] = sumwv;
	den[i] = sumw;
	rat[i] = sumwv/sumw;
      }
    }
  } else {
    /* general case */
    OUTERCHUNKLOOP(i, N, maxchunk, 16384) {
      R_CheckUserInterrupt();
      INNERCHUNKLOOP(i, N, maxchunk, 16384) {
	xi = x[i];
	yi = y[i];
	sumwv = sumw = 0.0;
	if(i > 0) {
	  for(j = 0; j < i; j++) {
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/pow(d2, pon2);
	    sumwv += w * v[j];
	    sumw  += w;
	  }
	}
	if(i < N-1) {
	  for(j = i+1; j < N; j++) {
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/pow(d2, pon2);
	    sumwv += w * v[j];
	    sumw  += w;
	  }
	}
	/* compute ratio */
	num[i] = sumwv;
	den[i] = sumw;
	rat[i] = sumwv/sumw;
      }
    }
  }
}

/* ----------------------------------------------------
   VERSIONS WITH VARIANCE CALCULATION
   --------------------------------------------------- */

/*  inverse-distance smoothing from data points onto pixel grid */

void Cidw2(x, y, v, n, xstart, xstep, nx, ystart, ystep, ny, power,
	   num, den, rat, mtwo, wtwo)
     double *x, *y, *v;           /* data points and values */
     int *n;
     double *xstart, *xstep, *ystart, *ystep;   /* pixel grid */
     int *nx, *ny;
     double *power;                   /* exponent for IDW */
     double *num, *den, *rat;    /* output arrays - assumed initialised 0 */
     double *mtwo, *wtwo;    /* output arrays - assumed initialised 0 */
{
  int N, i, Nx, Ny, ix, iy;
  double xg, yg, x0, dx, y0, dy, pon2, d2, w, vi,
    sumw, sumwv, sumw2, runmean, m2, delta, epsilon;
  
  N  = *n;
  Nx = *nx;
  Ny = *ny;
  x0 = *xstart;
  y0 = *ystart;
  dx = *xstep;
  dy = *ystep;

  pon2 = (*power)/2.0;

  if(pon2 == 1.0) {
    /* slightly faster code when power=2 */
    for(ix = 0, xg=x0; ix < Nx; ix++, xg+=dx) {
      if(ix % 256 == 0) R_CheckUserInterrupt();
      for(iy = 0, yg=y0; iy < Ny; iy++, yg+=dy) {
	sumwv = sumw = sumw2 = m2 = runmean = 0.0;
	/* loop over data points, accumulating numerator and denominator */
	for(i = 0; i < N; i++) {
	  vi = v[i];
	  d2 = (xg - x[i]) * (xg - x[i]) + (yg - y[i]) * (yg - y[i]);
	  w = 1.0/d2;
	  sumw  += w;
	  sumw2 += w * w;
	  sumwv += w * vi;
	  delta = vi - runmean;
	  epsilon = delta * w / sumw;
	  runmean += epsilon;
	  m2 += (sumw - w) * delta * epsilon;
	}
	/* compute ratio */
	MAT(num, ix, iy, Ny) = sumwv;
	MAT(den, ix, iy, Ny) = sumw;
	MAT(rat, ix, iy, Ny) = sumwv/sumw;
	MAT(mtwo, ix, iy, Ny) = m2;
	MAT(wtwo, ix, iy, Ny) = sumw2;
      }
    }
  } else {
    /* general case */
    for(ix = 0, xg=x0; ix < Nx; ix++, xg+=dx) {
      if(ix % 256 == 0) R_CheckUserInterrupt();
      for(iy = 0, yg=y0; iy < Ny; iy++, yg+=dy) {
	sumwv = sumw = sumw2 = m2 = runmean = 0.0;
	/* loop over data points, accumulating numerator and denominator */
	for(i = 0; i < N; i++) {
	  d2 = (xg - x[i]) * (xg - x[i]) + (yg - y[i]) * (yg - y[i]);
	  w = 1.0/pow(d2, pon2);
	  sumw  += w;
	  sumw2 += w * w;
	  sumwv += w * vi;
	  delta = vi - runmean;
	  epsilon = delta * w / sumw;
	  runmean += epsilon;
	  m2 += (sumw - w) * delta * epsilon;
	}
	/* compute ratio */
	MAT(num, ix, iy, Ny) = sumwv;
	MAT(den, ix, iy, Ny) = sumw;
	MAT(rat, ix, iy, Ny) = sumwv/sumw;
	MAT(mtwo, ix, iy, Ny) = m2;
	MAT(wtwo, ix, iy, Ny) = sumw2;
      }
    }
  }
}

/* Leave-one-out IDW at data points only */

void idwloo2(x, y, v, n, power, num, den, rat, mtwo, wtwo)
     double *x, *y, *v;           /* data points and values */
     int *n;
     double *power;                   /* exponent for IDW */
     double *num, *den, *rat, *mtwo, *wtwo;   /* output vectors - initialised 0 */
{
  int N, i, j, maxchunk;
  double xi, yi, d2, w, pon2,
    vj, sumw, sumwv, sumw2, runmean, m2, delta, epsilon;
  
  N  = *n;
  pon2 = (*power)/2.0;

  if(pon2 == 1.0) {
    /* slightly faster code when power=2 */
    OUTERCHUNKLOOP(i, N, maxchunk, 16384) {
      R_CheckUserInterrupt();
      INNERCHUNKLOOP(i, N, maxchunk, 16384) {
	xi = x[i];
	yi = y[i];
	sumwv = sumw = sumw2 = m2 = runmean = 0.0;
	if(i > 0) {
	  for(j = 0; j < i; j++) {
	    vj = v[j];
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/d2;
	    sumwv += w * vj;
	    sumw  += w;
	    sumw2 += w * w;
	    delta = vj - runmean;
	    epsilon = delta * w / sumw;
	    runmean += epsilon;
	    m2 += (sumw - w) * delta * epsilon;
	  }
	}
	if(i < N-1) {
	  for(j = i+1; j < N; j++) {
	    vj = v[j];
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/d2;
	    sumwv += w * vj;
	    sumw  += w;
	    sumw2 += w * w;
	    delta = vj - runmean;
	    epsilon = delta * w / sumw;
	    runmean += epsilon;
	    m2 += (sumw - w) * delta * epsilon;
	  }
	}
	/* compute ratio */
	num[i] = sumwv;
	den[i] = sumw;
	rat[i] = sumwv/sumw;
	mtwo[i] = m2;
	wtwo[i] = sumw2;
      }
    }
  } else {
    /* general case */
    OUTERCHUNKLOOP(i, N, maxchunk, 16384) {
      R_CheckUserInterrupt();
      INNERCHUNKLOOP(i, N, maxchunk, 16384) {
	xi = x[i];
	yi = y[i];
	sumwv = sumw = sumw2 = m2 = runmean = 0.0;
	if(i > 0) {
	  for(j = 0; j < i; j++) {
	    vj = v[j];
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/pow(d2, pon2);
	    sumwv += w * vj;
	    sumw  += w;
	    sumw2 += w * w;
	    delta = vj - runmean;
	    epsilon = delta * w / sumw;
	    runmean += epsilon;
	    m2 += (sumw - w) * delta * epsilon;
	  }
	}
	if(i < N-1) {
	  for(j = i+1; j < N; j++) {
	    vj = v[j];
	    d2 = (xi - x[j]) * (xi - x[j]) + (yi - y[j]) * (yi - y[j]);
	    w = 1.0/pow(d2, pon2);
	    sumwv += w * vj;
	    sumw  += w;
	    sumw2 += w * w;
	    delta = vj - runmean;
	    epsilon = delta * w / sumw;
	    runmean += epsilon;
	    m2 += (sumw - w) * delta * epsilon;
	  }
	}
	/* compute ratio */
	num[i] = sumwv;
	den[i] = sumw;
	rat[i] = sumwv/sumw;
	mtwo[i] = m2;
	wtwo[i] = sumw2;
      }
    }
  }
}

